flink-ml-iteration/src/main/java/org/apache/flink/iteration/operator/ReplayOperator.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.iteration.operator;
+
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.core.fs.FileSystem;
+import org.apache.flink.core.fs.Path;
+import org.apache.flink.iteration.IterationRecord;
+import org.apache.flink.iteration.config.IterationOptions;
+import org.apache.flink.iteration.datacache.nonkeyed.DataCacheReader;
+import org.apache.flink.iteration.datacache.nonkeyed.DataCacheWriter;
+import org.apache.flink.iteration.progresstrack.OperatorEpochWatermarkTracker;
+import org.apache.flink.iteration.progresstrack.OperatorEpochWatermarkTrackerFactory;
+import org.apache.flink.iteration.progresstrack.OperatorEpochWatermarkTrackerListener;
+import org.apache.flink.iteration.typeinfo.IterationRecordSerializer;
+import org.apache.flink.runtime.state.StateInitializationContext;
+import org.apache.flink.streaming.api.graph.StreamConfig;
+import org.apache.flink.streaming.api.operators.AbstractStreamOperator;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperator;
+import org.apache.flink.streaming.api.operators.Output;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.streaming.runtime.tasks.StreamTask;
+import org.apache.flink.util.ExceptionUtils;
+
+import java.io.IOException;
+import java.util.UUID;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.apache.flink.util.Preconditions.checkState;
+
+/** Replays the data received in the round 0 in the following round. */
+public class ReplayOperator<T> extends AbstractStreamOperator<IterationRecord<T>>
+        implements OneInputStreamOperator<IterationRecord<T>, IterationRecord<T>>,
+                OperatorEpochWatermarkTrackerListener {
+
+    private OperatorEpochWatermarkTracker progressTracker;
```

Comment:
I think it would work, but perhaps we may postpone the modification~? Since the current implementation satisfy the customs for `OneInputStreamOperator`

Reply:
I agree. But I think we might need a jira to track this improvement. WDYT?